### PR TITLE
[HIVEMALL-122] Added tokenize_cn UDF

### PR DESCRIPTION
**Pull request by partyyoung**
## What changes were proposed in this pull request?

Added tokenize_cn UDF to support word segmentation for Simplified Chinese text

## What type of PR is it?

Feature

## What is the Jira issue?

https://issues.apache.org/jira/browse/HIVEMALL-122

## How was this patch tested?

unit tests, manual tests

## How to use this feature?

udf

### CODE REVIEW

**Comment by myui** [comment]
Hi @partyyoung 

Thank you for your contribution!

Could you describe usage in https://github.com/apache/incubator-hivemall/blob/master/docs/gitbook/misc/tokenizer.md ?

`npm install -g gitbook-cli`
`cd docs/gitbook; gitbook` serve to check [the user guide](http://hivemall.incubator.apache.org/userguide/misc/tokenizer.html).

Also, `sec 2.3.` can be renamed as `Text Tokenizer`.

**Comment by myui** [comment]
Also, `tokenize_cn` should be added in [ddls](https://github.com/apache/incubator-hivemall/tree/master/resources/ddl).

grep `tokenize_ja` in the directory for reference.

**Comment by myui** [comment]
@partyyoung Thank you for fixing!

@takuti Could you review this PR and merge if it's looks good to you (for your experience as PPMC).

Here is a typical flow for merging a pull request to ASF git.
```
git checkout master
git checkout -b bug123
git pull http://repourl.git branch
git log | grep "Author" | head -1
git checkout master
git merge --squash bug123
git commit -a --author="Author <foo@me.com>" --message="Close #1: fixed a bug in xxx"
git push origin master
```

**Comment by takuti** [comment]
@myui sure!

File: resources/ddl/define-additional.hive [hunk 1]
```diff
@@ -9,6 +9,9 @@
 drop temporary function if exists tokenize_ja;
 create temporary function tokenize_ja as 'hivemall.nlp.tokenizer.KuromojiUDF';
 
+drop temporary function if exists tokenize_cn;
+create temporary function tokenize_cn as 'hivemall.nlp.tokenizer.SmartcnUDF';
```
Comment (takuti) [inline resources/ddl/define-additional.hive]: Could you also update `resources/ddl/define-udfs.td.hql`?

```
$ grep -r 'tokenize_ja' resources/ddl
resources/ddl/define-additional.hive:drop temporary function if exists tokenize_ja;
resources/ddl/define-additional.hive:create temporary function tokenize_ja as 'hivemall.nlp.tokenizer.KuromojiUDF';
resources/ddl/define-udfs.td.hql:create temporary function tokenize_ja as 'hivemall.nlp.tokenizer.KuromojiUDF';
```

File: nlp/src/test/java/hivemall/nlp/tokenizer/SmartcnUDFTest.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+		argOIs[1] = ObjectInspectorFactory
+				.getStandardConstantListObjectInspector(
+						PrimitiveObjectInspectorFactory.javaStringObjectInspector,
+						null);
+		udf.initialize(argOIs);
+		udf.close();
+	}
+
+	@Test
+	public void testEvalauteOneRow() throws IOException, HiveException {
```
Comment (takuti) [inline nlp/src/test/java/hivemall/nlp/tokenizer/SmartcnUDFTest.java]: Typo `Evalaute` => `Evaluate`

I know the typo originally comes from `KuromojiUDFTest.java` :) Could you fix both typos?

File: nlp/src/test/java/hivemall/nlp/tokenizer/SmartcnUDFTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorFactory;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory;
+import org.apache.hadoop.io.Text;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class SmartcnUDFTest {
+
+	@Test
+	public void testOneArgment() throws UDFArgumentException, IOException {
```
Comment (takuti) [inline nlp/src/test/java/hivemall/nlp/tokenizer/SmartcnUDFTest.java]: Could you fix typos `Argment` => `Argument` both in `KuromojiUDFTest` and `SmartcnUDFTest`?

File: docs/gitbook/misc/tokenizer.md [hunk 4]
```diff
@@ -46,4 +46,25 @@ select tokenize_ja("kuromojiを使った分かち書きのテストです。第
...
 
-For detailed APIs, please refer Javadoc of [JapaneseAnalyzer](https://lucene.apache.org/core/5_3_1/analyzers-kuromoji/org/apache/lucene/analysis/ja/JapaneseAnalyzer.html) as well.
\ No newline at end of file
+For detailed APIs, please refer Javadoc of [JapaneseAnalyzer](https://lucene.apache.org/core/5_3_1/analyzers-kuromoji/org/apache/lucene/analysis/ja/JapaneseAnalyzer.html) as well.
+
+# Tokenizer for Chinese Texts
+
+Hivemall-NLP module provides a Chinese text tokenizer UDF using [SmartChineseAnalyzer](http://lucene.apache.org/core/5_3_1/analyzers-smartcn/org/apache/lucene/analysis/cn/smart/SmartChineseAnalyzer.html). 
+
+> add jar /tmp/[hivemall-nlp-xxx-with-dependencies.jar](https://github.com/myui/hivemall/releases);
```
Comment (myui) [inline docs/gitbook/misc/tokenizer.md]: Also, better to remote the link to the old page `https://github.com/myui/hivemall/releases`.

You can remote the link in the `tokenize_ja` as well in this PR.

**Comment by partyyoung** [comment]
@takuti i am fixing these typos above. thank you.

**Comment by takuti** [comment]
@partyyoung Merged. Thank you for your contribution!

**Comment by myui** [comment]
It seems that code identification is strange in the latest commit. I'll fix it.
Better to apply `mvn formatter:format`.

**Comment by takuti** [comment]
@myui Yep, that's why I've created another PR #92. As long as coding guideline is not clearly specified, we should not force contributors to follow specific coding convention :)

**Comment by myui** [comment]
Apply `mvn formatter:format` should be commented in the contribution guide.

**Comment by takuti** [comment]
@myui It is! https://github.com/apache/incubator-hivemall/pull/92/files#diff-04c6e90faac2675aa89e2176d2eec7d8

**Pull request merged**